Crash in Extensions Search Results

### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

I'm not sure the full steps to reproduce, but I have my local Zed in a state after installing a dev extension where searching for the name of the dev extension (`ocaml`) will consistently crash the app.

### Zed Version and System Specs

Zed: v0.167.1 (Zed Preview)
OS: macOS 14.7.0
Memory: 64 GiB
Architecture: aarch64

### If applicable, add screenshots or screencasts of the incorrect state / behavior

_No response_

### If applicable, attach your Zed.log file to this issue.

```
2025-01-01T10:52:57.318703-05:00 [ERROR] {
  "thread": "main",
  "payload": "partition_at_index index 1 greater than length of slice 1",
  "location_data": {
    "file": "/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/slice/sort/select.rs",
    "line": 28
  },
  "backtrace": [
    "backtrace::capture::Backtrace::create::hbd8ec261cf178447+8777944",
    "backtrace::capture::Backtrace::new::h4d34e6163ddbe424+8777756",
    "zed::reliability::init_panic_hook::{{closure}}::h5c2c642c25421aae+77303060",
    "std::panicking::rust_panic_with_hook::h09e8a656f11e82b2+59018600",
    "std::panicking::begin_panic_handler::{{closure}}::h1230eb3cc91b241c+59015696",
    "std::sys::backtrace::__rust_end_short_backtrace::hc3491307aceda2c2+59008456",
    "_rust_begin_unwind+59014884",
    "core::panicking::panic_fmt::ha4b80a05b9fff47a+84736948",
    "fuzzy::strings::match_strings::{{closure}}::h05dd27a85be194b1+25773592",
    "extensions_ui::ExtensionsPage::fetch_extensions::{{closure}}::{{closure}}::hefd57fc1c2a36c41+25947156",
    "async_task::raw::RawTask<F,T,S,M>::run::h6c47996901d1886e+25395744",
    "gpui::platform::mac::dispatcher::trampoline::h596b636a83b1c6ee+30279888",
    "<gpui::platform::mac::platform::MacPlatform as gpui::platform::Platform>::run::hb57b3f8c2257fc2a+30340444",
    "gpui::app::App::run::h7ed3cd68b78cabe0+75546048",
    "zed::main::h3d01cc3a7aeb1588+77347156",
    "std::sys::backtrace::__rust_begin_short_backtrace::h21e917c7877de9d9+77203468",
    "std::rt::lang_start::{{closure}}::h6a2316a6f53fd709+77139516",
    "std::rt::lang_start_internal::hdd117cb81a316264+58950236",
    "_main+77359000"
  ],
  "app_version": "0.168.0",
  "release_channel": "nightly",
  "target": "aarch64-apple-darwin",
  "os_name": "macOS",
  "os_version": "14.7.0",
  "architecture": "aarch64",
  "panicked_on": 1735746777318,
  "system_id": "bc9bfc98-cb59-4b78-bfc8-2ad1d877d7e1",
  "installation_id": "048be47f-8cfd-498f-8491-5c7b7ae260ad",
  "session_id": "4ad2858d-a17a-48fa-8915-32bdb3ca70ca"
}
```

--- a/crates/fuzzy/src/matcher.rs
+++ b/crates/fuzzy/src/matcher.rs
@@ -14,11 +14,17 @@ pub struct Matcher<'a> {
     lowercase_query: &'a [char],
     query_char_bag: CharBag,
     smart_case: bool,
+    max_results: usize,
     min_score: f64,
     match_positions: Vec<usize>,
     last_positions: Vec<usize>,
     score_matrix: Vec<Option<f64>>,
     best_position_matrix: Vec<usize>,
+}
+
+pub trait Match: Ord {
+    fn score(&self) -> f64;
+    fn set_positions(&mut self, positions: Vec<usize>);
 }
 
 pub trait MatchCandidate {
@@ -32,6 +38,7 @@ impl<'a> Matcher<'a> {
         lowercase_query: &'a [char],
         query_char_bag: CharBag,
         smart_case: bool,
+        max_results: usize,
     ) -> Self {
         Self {
             query,
@@ -43,11 +50,10 @@ impl<'a> Matcher<'a> {
             score_matrix: Vec::new(),
             best_position_matrix: Vec::new(),
             smart_case,
+            max_results,
         }
     }
 
-    /// Filter and score fuzzy match candidates. Results are returned unsorted, in the same order as
-    /// the input candidates.
     pub fn match_candidates<C: MatchCandidate, R, F>(
         &mut self,
         prefix: &[char],
@@ -57,7 +63,8 @@ impl<'a> Matcher<'a> {
         cancel_flag: &AtomicBool,
         build_match: F,
     ) where
-        F: Fn(&C, f64, &Vec<usize>) -> R,
+        R: Match,
+        F: Fn(&C, f64) -> R,
     {
         let mut candidate_chars = Vec::new();
         let mut lowercase_candidate_chars = Vec::new();
@@ -96,7 +103,20 @@ impl<'a> Matcher<'a> {
             );
 
             if score > 0.0 {
-                results.push(build_match(&candidate, score, &self.match_positions));
+                let mut mat = build_match(&candidate, score);
+                if let Err(i) = results.binary_search_by(|m| mat.cmp(m)) {
+                    if results.len() < self.max_results {
+                        mat.set_positions(self.match_positions.clone());
+                        results.insert(i, mat);
+                    } else if i < results.len() {
+                        results.pop();
+                        mat.set_positions(self.match_positions.clone());
+                        results.insert(i, mat);
+                    }
+                    if results.len() == self.max_results {
+                        self.min_score = results.last().unwrap().score();
+                    }
+                }
             }
         }
     }
@@ -305,18 +325,18 @@ mod tests {
     #[test]
     fn test_get_last_positions() {
         let mut query: &[char] = &['d', 'c'];
-        let mut matcher = Matcher::new(query, query, query.into(), false);
+        let mut matcher = Matcher::new(query, query, query.into(), false, 10);
         let result = matcher.find_last_positions(&['a', 'b', 'c'], &['b', 'd', 'e', 'f']);
         assert!(!result);
 
         query = &['c', 'd'];
-        let mut matcher = Matcher::new(query, query, query.into(), false);
+        let mut matcher = Matcher::new(query, query, query.into(), false, 10);
         let result = matcher.find_last_positions(&['a', 'b', 'c'], &['b', 'd', 'e', 'f']);
         assert!(result);
         assert_eq!(matcher.last_positions, vec![2, 4]);
 
         query = &['z', '/', 'z', 'f'];
-        let mut matcher = Matcher::new(query, query, query.into(), false);
+        let mut matcher = Matcher::new(query, query, query.into(), false, 10);
         let result = matcher.find_last_positions(&['z', 'e', 'd', '/'], &['z', 'e', 'd', '/', 'f']);
         assert!(result);
         assert_eq!(matcher.last_positions, vec![0, 3, 4, 8]);
@@ -431,7 +451,7 @@ mod tests {
             });
         }
 
-        let mut matcher = Matcher::new(&query, &lowercase_query, query_chars, smart_case);
+        let mut matcher = Matcher::new(&query, &lowercase_query, query_chars, smart_case, 100);
 
         let cancel_flag = AtomicBool::new(false);
         let mut results = Vec::new();
@@ -442,17 +462,16 @@ mod tests {
             path_entries.into_iter(),
             &mut results,
             &cancel_flag,
-            |candidate, score, positions| PathMatch {
+            |candidate, score| PathMatch {
                 score,
                 worktree_id: 0,
-                positions: positions.clone(),
+                positions: Vec::new(),
                 path: Arc::from(candidate.path),
                 path_prefix: "".into(),
                 distance_to_relative_ancestor: usize::MAX,
                 is_dir: false,
             },
         );
-        results.sort_by(|a, b| b.cmp(a));
 
         results
             .into_iter()

--- a/crates/fuzzy/src/paths.rs
+++ b/crates/fuzzy/src/paths.rs
@@ -3,10 +3,7 @@ use std::{
     borrow::Cow,
     cmp::{self, Ordering},
     path::Path,
-    sync::{
-        atomic::{self, AtomicBool},
-        Arc,
-    },
+    sync::{atomic::AtomicBool, Arc},
 };
 
 use crate::{
@@ -157,10 +154,6 @@ pub async fn match_path_sets<'a, Set: PathMatchCandidateSet<'a>>(
 
                     let mut tree_start = 0;
                     for candidate_set in candidate_sets {
-                        if cancel_flag.load(atomic::Ordering::Relaxed) {
-                            break;
-                        }
-
                         let tree_end = tree_start + candidate_set.len();
 
                         if tree_start < segment_end && segment_start < tree_end {
@@ -208,10 +201,6 @@ pub async fn match_path_sets<'a, Set: PathMatchCandidateSet<'a>>(
             }
         })
         .await;
-
-    if cancel_flag.load(atomic::Ordering::Relaxed) {
-        return Vec::new();
-    }
 
     let mut results = segment_results.concat();
     util::truncate_to_bottom_n_sorted_by(&mut results, max_results, &|a, b| b.cmp(a));

--- a/crates/fuzzy/src/paths.rs
+++ b/crates/fuzzy/src/paths.rs
@@ -7,7 +7,7 @@ use std::{
 };
 
 use crate::{
-    matcher::{MatchCandidate, Matcher},
+    matcher::{Match, MatchCandidate, Matcher},
     CharBag,
 };
 
@@ -40,6 +40,16 @@ pub trait PathMatchCandidateSet<'a>: Send + Sync {
     }
     fn prefix(&self) -> Arc<str>;
     fn candidates(&'a self, start: usize) -> Self::Candidates;
+}
+
+impl Match for PathMatch {
+    fn score(&self) -> f64 {
+        self.score
+    }
+
+    fn set_positions(&mut self, positions: Vec<usize>) {
+        self.positions = positions;
+    }
 }
 
 impl<'a> MatchCandidate for PathMatchCandidate<'a> {
@@ -92,7 +102,13 @@ pub fn match_fixed_path_set(
     let query = query.chars().collect::<Vec<_>>();
     let query_char_bag = CharBag::from(&lowercase_query[..]);
 
-    let mut matcher = Matcher::new(&query, &lowercase_query, query_char_bag, smart_case);
+    let mut matcher = Matcher::new(
+        &query,
+        &lowercase_query,
+        query_char_bag,
+        smart_case,
+        max_results,
+    );
 
     let mut results = Vec::new();
     matcher.match_candidates(
@@ -101,17 +117,16 @@ pub fn match_fixed_path_set(
         candidates.into_iter(),
         &mut results,
         &AtomicBool::new(false),
-        |candidate, score, positions| PathMatch {
+        |candidate, score| PathMatch {
             score,
             worktree_id,
-            positions: positions.clone(),
+            positions: Vec::new(),
             is_dir: candidate.is_dir,
             path: Arc::from(candidate.path),
             path_prefix: Arc::default(),
             distance_to_relative_ancestor: usize::MAX,
         },
     );
-    util::truncate_to_bottom_n_sorted_by(&mut results, max_results, &|a, b| b.cmp(a));
     results
 }
 
@@ -149,8 +164,13 @@ pub async fn match_path_sets<'a, Set: PathMatchCandidateSet<'a>>(
                 scope.spawn(async move {
                     let segment_start = segment_idx * segment_size;
                     let segment_end = segment_start + segment_size;
-                    let mut matcher =
-                        Matcher::new(query, lowercase_query, query_char_bag, smart_case);
+                    let mut matcher = Matcher::new(
+                        query,
+                        lowercase_query,
+                        query_char_bag,
+                        smart_case,
+                        max_results,
+                    );
 
                     let mut tree_start = 0;
                     for candidate_set in candidate_sets {
@@ -173,10 +193,10 @@ pub async fn match_path_sets<'a, Set: PathMatchCandidateSet<'a>>(
                                 candidates,
                                 results,
                                 cancel_flag,
-                                |candidate, score, positions| PathMatch {
+                                |candidate, score| PathMatch {
                                     score,
                                     worktree_id,
-                                    positions: positions.clone(),
+                                    positions: Vec::new(),
                                     path: Arc::from(candidate.path),
                                     is_dir: candidate.is_dir,
                                     path_prefix: candidate_set.prefix(),
@@ -202,8 +222,14 @@ pub async fn match_path_sets<'a, Set: PathMatchCandidateSet<'a>>(
         })
         .await;
 
-    let mut results = segment_results.concat();
-    util::truncate_to_bottom_n_sorted_by(&mut results, max_results, &|a, b| b.cmp(a));
+    let mut results = Vec::new();
+    for segment_result in segment_results {
+        if results.is_empty() {
+            results = segment_result;
+        } else {
+            util::extend_sorted(&mut results, segment_result, max_results, |a, b| b.cmp(a));
+        }
+    }
     results
 }
 

--- a/crates/fuzzy/src/strings.rs
+++ b/crates/fuzzy/src/strings.rs
@@ -8,7 +8,7 @@ use std::{
     cmp::{self, Ordering},
     iter,
     ops::Range,
-    sync::atomic::{self, AtomicBool},
+    sync::atomic::AtomicBool,
 };
 
 #[derive(Clone, Debug)]
@@ -177,10 +177,6 @@ pub async fn match_strings(
             }
         })
         .await;
-
-    if cancel_flag.load(atomic::Ordering::Relaxed) {
-        return Vec::new();
-    }
 
     let mut results = segment_results.concat();
     util::truncate_to_bottom_n_sorted_by(&mut results, max_results, &|a, b| b.cmp(a));

--- a/crates/fuzzy/src/strings.rs
+++ b/crates/fuzzy/src/strings.rs
@@ -1,5 +1,5 @@
 use crate::{
-    matcher::{MatchCandidate, Matcher},
+    matcher::{Match, MatchCandidate, Matcher},
     CharBag,
 };
 use gpui::BackgroundExecutor;
@@ -44,6 +44,16 @@ pub struct StringMatch {
     pub score: f64,
     pub positions: Vec<usize>,
     pub string: String,
+}
+
+impl Match for StringMatch {
+    fn score(&self) -> f64 {
+        self.score
+    }
+
+    fn set_positions(&mut self, positions: Vec<usize>) {
+        self.positions = positions;
+    }
 }
 
 impl StringMatch {
@@ -157,8 +167,13 @@ pub async fn match_strings(
                 scope.spawn(async move {
                     let segment_start = cmp::min(segment_idx * segment_size, candidates.len());
                     let segment_end = cmp::min(segment_start + segment_size, candidates.len());
-                    let mut matcher =
-                        Matcher::new(query, lowercase_query, query_char_bag, smart_case);
+                    let mut matcher = Matcher::new(
+                        query,
+                        lowercase_query,
+                        query_char_bag,
+                        smart_case,
+                        max_results,
+                    );
 
                     matcher.match_candidates(
                         &[],
@@ -166,10 +181,10 @@ pub async fn match_strings(
                         candidates[segment_start..segment_end].iter(),
                         results,
                         cancel_flag,
-                        |candidate, score, positions| StringMatch {
+                        |candidate, score| StringMatch {
                             candidate_id: candidate.id,
                             score,
-                            positions: positions.clone(),
+                            positions: Vec::new(),
                             string: candidate.string.to_string(),
                         },
                     );
@@ -178,7 +193,13 @@ pub async fn match_strings(
         })
         .await;
 
-    let mut results = segment_results.concat();
-    util::truncate_to_bottom_n_sorted_by(&mut results, max_results, &|a, b| b.cmp(a));
+    let mut results = Vec::new();
+    for segment_result in segment_results {
+        if results.is_empty() {
+            results = segment_result;
+        } else {
+            util::extend_sorted(&mut results, segment_result, max_results, |a, b| b.cmp(a));
+        }
+    }
     results
 }

--- a/crates/util/src/util.rs
+++ b/crates/util/src/util.rs
@@ -8,6 +8,7 @@ pub mod test;
 
 use anyhow::{anyhow, Context as _, Result};
 use futures::Future;
+
 use itertools::Either;
 use regex::Regex;
 use std::sync::{LazyLock, OnceLock};
@@ -108,27 +109,6 @@ where
             start_index = index;
         }
     }
-}
-
-pub fn truncate_to_bottom_n_sorted_by<T, F>(items: &mut Vec<T>, limit: usize, compare: &F)
-where
-    F: Fn(&T, &T) -> Ordering,
-{
-    if limit == 0 {
-        items.truncate(0);
-    }
-    if items.len() < limit {
-        return;
-    }
-    // When limit is near to items.len() it may be more efficient to sort the whole list and
-    // truncate, rather than always doing selection first as is done below. It's hard to analyze
-    // where the threshold for this should be since the quickselect style algorithm used by
-    // `select_nth_unstable_by` makes the prefix partially sorted, and so its work is not wasted -
-    // the expected number of comparisons needed by `sort_by` is less than it is for some arbitrary
-    // unsorted input.
-    items.select_nth_unstable_by(limit, compare);
-    items.truncate(limit);
-    items.sort_by(compare);
 }
 
 #[cfg(unix)]

